Code/Framework/AzCore/AzCore/Instance/InstancePool.h

```diff
@@ -108,8 +117,16 @@ namespace AZ
                 else
                 {
                     m_nameToInstancePool.erase(insertResult.first);
-                    m_nameToInstancePool.insert({ name, instancePool });
-                    return AZ::Success(instancePool);
+                    insertResult = m_nameToInstancePool.insert({ name, instancePool });
+
+                    if (insertResult.second)
+                    {
+                        return AZ::Success(instancePool);
+                    }
+                    else
+                    {
+                        return AZ::Failure(AZStd::string::format("Could not add InstancePool with name %.s", name.GetCStr()));
+                    }
```

Comment:
Is that the only reason that an insert can fail? What about out of memory, throw during construction, etc? Your solution would assert (which currently prints a message to the console and does *not* break by default) then return Success, even on failure. The current solution uses our failure mechanism to pass a coherent message, and the both have the same number of checks. I don't see the advantage in changing this.

Reply:
The engine would be bloated with conditional branching if it cannot be trusted that a simple `insert` would work as expected, and performance would suffer. It wouldn't be possible to do a single operation without having to check an error for it. It's great that you care for problems at that deep level, but in my opinion for this case it might be too much, an assert is good enough to tell that something unexpected has happened.

Reply:
how does an assert skip conditional branching? It has to test the assert condition, correct? It's like, it's only job.

Reply:
Bit late on this one, but wanted to say, the assert would skip conditional branching in a release build, but we don't emit release builds (to my knowledge) for working in the editor, we use profile builds which have asserts enabled. If this code is to be used in performance critical code (i.e. in a runtime game environment) then I agree the assert may be more appropriate. If this is editor/tool code, then perhaps this approach is reasonable.